src/Evolution/Particles/MonteCarlo/Actions/FluidCouplingAction.hpp

```diff
@@ -0,0 +1,122 @@
+    get(det_jacobian_logical_to_inertial) =
+        1.0 / get(det_inverse_jacobian_logical_to_inertial);
+    Scalar<DataVector> cell_inertial_three_volume =
+        make_with_value<Scalar<DataVector>>(*tilde_tau, 0.0);
+    cell_inertial_coordinate_three_volume_finite_difference(
+        &cell_inertial_three_volume, mesh, det_jacobian_logical_to_inertial);
+
+    for (size_t i = 0; i < extents[0]; i++) {
+      for (size_t j = 0; j < extents[1]; j++) {
+        for (size_t k = 0; k < extents[2]; k++) {
+          // The coupling terms are computed on a grid with ghost zone points,
+          // the fluid variables are on the grid without GZ points
+          // Index without GZ
+          const size_t local_idx =
+              i + j * extents[0] + k * extents[0] * extents[1];
+          // Index with GZ
+          const size_t extended_idx =
+              num_ghost_zones + i +
+              (num_ghost_zones + j) * (extents[0] + 2 * num_ghost_zones) +
+              (num_ghost_zones + k) * (extents[0] + 2 * num_ghost_zones) *
+                  (extents[1] + 2 * num_ghost_zones);
+          // The MC coupling calculates the change in energy, momentum, and
+          // lepton number. We need to divide by the cell 3-volume to get
+          // the change in the evolved variables.
+          const double& volume = get(cell_inertial_three_volume)[local_idx];
+          get(*tilde_tau)[local_idx] +=
+              get(*coupling_tilde_tau)[extended_idx] / volume;
+          get(*tilde_ye)[local_idx] +=
+              get(*coupling_tilde_rho_ye)[extended_idx] / volume;
+          for (size_t d = 0; d < 3; d++) {
+            tilde_s->get(d)[local_idx] +=
+                coupling_tilde_s->get(d)[extended_idx] / volume;
+          }
+        }
+      }
+    }
+    // Reset coupling terms to 0 after use
+    get(*coupling_tilde_tau) = 0.0;
+    get(*coupling_tilde_rho_ye) = 0.0;
+    for (size_t d = 0; d < 3; d++) {
+      coupling_tilde_s->get(d) = 0.0;
+    }
+  }
+};
+
+namespace Actions {
+
+/// Action coupling the Monte-Carlo packets to the fluid
+struct CoupleMonteCarloToFluid {
+  template <typename DbTags, typename... InboxTags, typename ArrayIndex,
+            typename ActionList, typename ParallelComponent,
+            typename Metavariables>
+  static Parallel::iterable_action_return_t apply(
+      db::DataBox<DbTags>& box, tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) {
+    ASSERT(db::get<evolution::dg::subcell::Tags::ActiveGrid>(box) ==
+               evolution::dg::subcell::ActiveGrid::Subcell,
+           "MC assumes that we are using the Subcell grid!");
+
+    db::mutate_apply(FluidCouplingMutator{}, make_not_null(&box));
+    return {Parallel::AlgorithmExecution::Continue, std::nullopt};
+  }
+};
```

Comment:
You can use the action `::Actions::MutateApply<Particles::MonteCarlo::FluidCouplingMutator>` where the action is in `src/ParallelAlgorithms/Actions/MutateApply.hpp`, then you don't need the action you have here at all. You could move the `ASSERT` to the mutator as well.